File: Google/1153. String Transforms Into Another String.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {boolean}
 */
var canConvert = function(str1, str2) {
    if (!str1 || !str2 || str1.length != str2.length) return false;

    let map = new Map();
    let changed = 0;
    let charsSet = new Set();
    let swap = false;
    for (let i = 0; i < str1.length; i++) {
        const convertFrom = str1[i];
        const convertTo = str2[i];

        if (!map.has(convertFrom)) {
            map.set(convertFrom, convertTo);

            if (!charsSet.has(convertTo)) charsSet.add(convertTo);

            if (convertFrom !== convertTo) {
                changed++;

                if (!swap && map.has(convertTo) && map.get(convertTo) === convertFrom)
                    swap = true;
            }
        } else if (map.get(convertFrom) !== convertTo) {
            return false;
        }
    }

    return charsSet.size < 26 || (!swap && changed < 26);
};

canConvert("abcdefghijklmnopqrstuvwxyz", "abcdefghijklmnopqrstuvwxzy");

File: Google/1153. String Transforms Into Another String.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LeetCode
{
    public class Solution
    {
        public bool CanConvert(string str1, string str2)
        {
            if (string.IsNullOrEmpty(str1) || string.IsNullOrEmpty(str2) || str1.Length != str2.Length) return false;

            var map = new Dictionary<char, char>();
            var changed = 0;
            var swap = false;
            var chars = new HashSet<char>();
            for (var i = 0; i < str1.Length; i++)
            {
                var key = str1[i];
                var value = str2[i];

                if (!map.ContainsKey(key))
                {
                    map.Add(key, value);

                    if (key != value)
                    {
                        changed++;
                        if (!swap && map.ContainsKey(value) && map[value] == key) swap = true;
                    }
                }
                else if (map[key] != value)
                {
                    return false;
                }

                if (!chars.Contains(value)) chars.Add(value);
            }

            return chars.Count < 26 || (changed < 26 && !swap);
        }
    }
}
